### PR TITLE
rework error system a bit to remove a bunch of duplicated code

### PR DESCRIPTION
**Pull request by justcool393**
one of the first things i did is do an errors rework over on rdrama. this uses flask's simple and relatively stable way to abort out of a function quickly everywhere, and also we can provide extra details to the user

for example

image

this message also is displayed in the API as well

see also:
https://github.com/Aevann1/rDrama/pull/400
https://github.com/justcool393/rDrama/tree/errors-rework

this also adds the 409 and 422 code (i'm amenable to changing the message)
this should prolly be squashed and merged

### CODE REVIEW

**Comment by justcool393** [comment]
looks to be good. have tested it and it should work pretty well

File: files/routes/posts.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 from files.helpers.sanitize import *
 from files.helpers.strings import sql_ilike_clean
 from files.helpers.alerts import *
+from files.helpers.discord import send_discord_message, send_cringetopia_message
```
Comment (TLSM) [inline files/routes/posts.py]: You accidentally a line.

File: files/routes/posts.py [hunk 2]
```diff
@@ -105,6 +105,7 @@ def publish(pid, v):
 	cache.delete_memoized(User.userpagelisting)
 
 	if v.admin_level > 0 and ("[changelog]" in post.title.lower() or "(changelog)" in post.title.lower()):
+		send_discord_message(post.permalink)
```
Comment (TLSM) [inline files/routes/posts.py]: A line.

File: files/routes/posts.py [hunk 3]
```diff
@@ -1005,6 +998,7 @@ def error(error):
 	cache.delete_memoized(User.userpagelisting)
 
 	if v.admin_level > 0 and ("[changelog]" in post.title.lower() or "(changelog)" in post.title.lower()) and not post.private:
+		send_discord_message(post.permalink)
```
Comment (TLSM) [inline files/routes/posts.py]: ... a line.

File: files/routes/posts.py [hunk 4]
```diff
@@ -1125,35 +1119,39 @@ def unsave_post(pid, v):
...
+	else: return {"message": "Post unpinned!"}
 
 @app.get("/submit/title")
 @limiter.limit("6/minute")
 @auth_required
 def get_post_title(v):
-
 	url = request.values.get("url")
-	if not url: abort(400)
+	if not url or '\\' in url: abort(400)
```
Comment (TLSM) [inline files/routes/posts.py:1136]: Merge up?

**Review by TLSM — CHANGES_REQUESTED** [review]

File: files/routes/posts.py (hunk 4 above)
Comment (justcool393) [reply files/routes/posts.py:1136]: again not sure how this got in here but... sure why not

**Comment by justcool393** [comment]
ty
if you could also just spot check my merge too i'd rly appreciate it

**Review by TLSM — APPROVED** [review]
Tested locally:
  - Server starts, viewing common pages, submitting new posts, and commenting all still work.
  - cURL request while bots disabled gets a properly formed 403 JSON response body.
  - Error toast correctly displays in UI on comment 403.
  - Submitting a post with empty title (after bypassing frontend verification) correctly gets a 400 page "Minimum length for title is 1".

Also we have very similar errors.py logic on the upstream already which has been stable for a number of weeks. If there's any bug here, it'd likely be localized to a particular `abort(...)` location.

**Comment by zorbathut** [comment]
Lookin' good!

If at any point you can deduplicate some of that

```
if (data && data["error"]) document.getElementById('toast-post-error-text').innerText = data["error"];
if (data && data["details"]) document.getElementById('toast-post-error-text').innerText = data["details"];
```

stuff, that would also be great :V

**Comment by justcool393** [comment]
agreed. will put it on a todo. it's mostly a weird artifiact of this fork having like 5 posttoast variations

**Pull request merged**